File: leetcode.com/0097_interleaving_string/solution.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:

        @cache
        def dfs(i, j, k):
            if k == len(s3):
                return True

            if i < len(s1) and s1[i] == s3[k] and dfs(i + 1, j, k + 1):
                return True

            if j < len(s2) and s2[j] == s3[k] and dfs(i, j + 1, k + 1):
                return True

            return False

        if len(s1) + len(s2) != len(s3):
            return False

        return dfs(0, 0, 0)

File: leetcode.com/0097_interleaving_string/solution.c
char *p1, *p2, *p3;
int l1, l2;

int cache[101][101];

bool dfs(int i, int j)
{
    if (cache[i][j] != -1)
        return cache[i][j];

    if (i == l1 && j == l2)
        return true;

    bool r1 = false;
    bool r2 = false;

    if (i < l1 && p1[i] == p3[i+j])
        r1 = dfs(i + 1, j);

    if (j < l2 && p2[j] == p3[i+j])
        r2 = dfs(i, j + 1);

    cache[i][j] = r1 || r2;
    return cache[i][j];
}

bool isInterleave(char *s1, char *s2, char *s3)
{
    l1 = strlen(s1);
    l2 = strlen(s2);

    if (l1 + l2 != strlen(s3))
        return false;

    p1 = s1; p2 = s2; p3 = s3;
    memset(cache, -1, sizeof(cache));
    return dfs(0, 0);
}
